File: src/cpp/Brian/Console.cpp
//----------------------------------------------------------------------------
//
//       Copyright (c) 2021-2025 Frank Eskesen.
//
//       This file is free content, distributed under the GNU General
//       Public License, version 3.0.
//       (See accompanying file LICENSE.GPL-3.0 or the original
//       contained within https://www.gnu.org/licenses/gpl-3.0.en.html)
//
// SPDX-License-Identifier: GPL-3.0-only
//----------------------------------------------------------------------------
//
// Title-
//       Console.cpp
//
// Purpose-
//       Operate the input terminal
//
// Last change date-
//       2025/03/01
//
// Implementation note-
//       When running using a static library build, HCDM debugging displays in
//       start() and run() should be disabled. The debug RecursiveLatch unlock
//       may fail with a terminating error.
//       (This should not occur, but has not been debugged.)
//
//----------------------------------------------------------------------------
#include <mutex>                    // For std::mutex, std::lock_guard
#include <cctype>                   // For isspace
#include <cstring>                  // For strlen

#include <unistd.h>                 // For isatty, STDIN_FILENO, ...

#include <pub/Console.h>            // For pub::Console
#include <pub/Debug.h>              // For debugging
#include <pub/Thread.h>             // For pub::Thread

#include "Command.h"                // For class Command
#include "Common.h"                 // For class Common
#include "Console.h"                // For class CommonThread, CommonService
#include "Service.h"                // For class Service

#define PUB _LIBPUB_NAMESPACE
using namespace PUB::debugging;     // For debugging subroutines

using PUB::Debug;                   // For class pub::Debug
using PUB::Thread;                  // For class pub::Thread

//----------------------------------------------------------------------------
// Constants for parameterization
//----------------------------------------------------------------------------
enum
{  HCDM= false                      // Hard Core Debug Mode?
,  VERBOSE= 0                       // Verbosity, higher is more verbose

,  USE_COMMAND_ECHOING= true        // Echo commands to trace file?
}; // (generic) enum

//----------------------------------------------------------------------------
// External data areas
//----------------------------------------------------------------------------
ConsoleService         consoleService; // The ConsoleService

//----------------------------------------------------------------------------
// Internal data areas
//----------------------------------------------------------------------------
std::mutex             mutex;       // (Hidden) mutex

//----------------------------------------------------------------------------
//
// Subroutine-
//       strip
//
// Purpose-
//       Strip leading and trailing whitespace.
//
//----------------------------------------------------------------------------
static char*                        // The stripped string
   strip(                           // Strip leading and trailing whitespace
     char*             C)           // From this string (MODIFIED)
{
   while( isspace(*C) )             // Strip leading whitespace
     ++C;

   int L= strlen((char*)C);
   while( L > 0 && isspace(C[L-1]) ) // Strip trailing whitespace
     C[--L]= '\0';

   return C;
}

//----------------------------------------------------------------------------
//
// Methods-
//       ConsoleThread::ConsoleThread
//       ConsoleThread::~ConsoleThread
//
// Purpose-
//       Constructor.
//       Destructor.
//
//----------------------------------------------------------------------------
   ConsoleThread::ConsoleThread( void ) // Constructor
:  pub::Thread()
{  if( HCDM ) debugh("ConsoleThread(%p).!\n", this);

   if( !isatty(STDIN_FILENO) || !isatty(STDOUT_FILENO) ) {
     errorf("ERROR: ConsoleThread only supports terminal input/output\n");
     exit(1);
   }
}

   ConsoleThread::~ConsoleThread( void ) // Destructor
{  if( HCDM ) debugh("ConsoleThread(%p).~\n", this); }

//----------------------------------------------------------------------------
// ConsoleThread::Methods
//----------------------------------------------------------------------------
int
   ConsoleThread::getch( void )     // Get character from stdin
{  return pub::Console::getch(); }

void
   ConsoleThread::putch(            // Put character onto stdout
     int               C)           // The character
{  pub::Console::putch(C); }

char*                               // The input line
   ConsoleThread::readline( void )  // Read input line
{
   if( USE_COMMAND_ECHOING ) {
     tracef("\n");
   }
   pub::Console::puts(">>> ");           // Input prompt
   pub::Console::gets(inp, sizeof(inp)-1);
   char* C= strip(inp);
   if( USE_COMMAND_ECHOING ) {
     traceh("==> '%s'\n", C);
   }
   return C;
}

void
   ConsoleThread::run( void )       // The operational thread
{  if( HCDM ) debugh("ConsoleThread(%p).run\n", this);

   pub::Console::start();
   if( HCDM )
     debugh("pub::Console::started\n");

   operational= true;
   sleep(1);                        // One second startup delay

   while( operational ) {
     char* C= readline();
     if( operational )
       Command::command(C);         // Run the command, ignoring any resultant

   }

   if( HCDM && VERBOSE > 0 ) debugh("pub::Console::run complete\n");
}

void
   ConsoleThread::stop( void )      // Terminate the thread
{  if( HCDM ) debugh("ConsoleThread(%p).stop\n", this);

   if( operational )
     pub::Console::stop();

   operational= false;
}

void
   ConsoleThread::wait( void )      // Wait for termination completion
{  if( HCDM ) debugh("ConsoleThread(%p).wait\n", this);

   pub::Console::wait();            // Wait for the Console
}

//----------------------------------------------------------------------------
//
// Methods-
//       ConsoleService::ConsoleService
//       ConsoleService::~ConsoleService
//
// Purpose-
//       Constructors
//       Destructors
//
//----------------------------------------------------------------------------
   ConsoleService::ConsoleService( void )      // Constructor
:  Service("Console") {}

   ConsoleService::~ConsoleService( void )     // Destructor
{
   std::lock_guard<decltype(mutex)> lock(mutex);
   delete console_thread;
   console_thread= nullptr;
}

//----------------------------------------------------------------------------
// ConsoleService::Methods
//----------------------------------------------------------------------------
void
   ConsoleService::start(Service* S) // Start the ConsoleService
{  if( HCDM ) debugh("ConsoleService(%p).start(%p)\n", this, S);
   Service::has_start::start(this);

   std::lock_guard<decltype(mutex)> lock(mutex);
   if( console_thread ) {
     debugh("ConsoleService::start ERROR: already started\n");
     return;
   }

   console_thread= new ConsoleThread();
   console_thread->start();
}

void
   ConsoleService::stop(Service*)   // Stop the ConsoleService
{  if( HCDM ) debugh("ConsoleService(%p).stop\n", this);
   Service::has_stop::stop(this);

   std::lock_guard<decltype(mutex)> lock(mutex);
   if( console_thread )
     console_thread->stop();        // Stop the ConsoleThread
}

void
   ConsoleService::wait(Service*)   // Wait for ConsoleService termination
{  if( HCDM ) debugh("ConsoleService(%p).wait\n", this);
   Service::has_wait::wait(this);

   std::lock_guard<decltype(mutex)> lock(mutex);
   if( console_thread ) {
     console_thread->wait();        // Wait for the ConsoleThread
     console_thread->join();        // Join (complete) the ConsoleThread

     delete console_thread;
     console_thread= nullptr;
   }
}

//----------------------------------------------------------------------------
//
// Class-
//       Command_quit
//
// Purpose-
//       Terminate processing
//
//----------------------------------------------------------------------------
class Command_quit : public Command {
public:
   Command_quit() : Command("quit")
{  }

virtual Command::resultant          // Resultant
   main(int, char**)                // Handle Command
{
   {{{{
     std::lock_guard<decltype(mutex)> lock(mutex);
     if( consoleService.console_thread )
       consoleService.console_thread->stop();
   }}}}

   Common::get()->shutdown();
   return nullptr;
}
}  command_quit; // static class Command_quit

File: src/py/Brian/Console.py
##############################################################################
##
##       Copyright (C) 2016-2018 Frank Eskesen.
##
##       This file is free content, distributed under the GNU General
##       Public License, version 3.0.
##       (See accompanying file LICENSE.GPL-3.0 or the original
##       contained within https://www.gnu.org/licenses/gpl-3.0.en.html)
##
##############################################################################
##
## Title-
##       Console.py
##
## Purpose-
##       Brian AI: Manage operator console
##
## Last change date-
##       2018/01/01
##
## Usage notes[backspace handling]-
##       The Console should be run from a terminal which handles the backspace
##       character itself, so that readline does not return it.
##
##       If it isn't, sys.stdin.readline echos the backspace and passes it
##       through to be interpreted. While we correctly remove these backspaces
##       the resulting line will not be displayed correctly if that echoing
##       overwrites the prompt.
##
##       The stdio assembler package provides versions of getch and putch
##       which do not present this problem. The getch method does not echo
##       and all characters, including backspaces, are returned as received.
##       We don't echo backspaces past the original position.
##
## Implementation notes-
##       Neither sys.stdin.readline nor sys.stdin.read(1) returns any input
##       until an end-of-line  character is received. Both of these routines
##       also handle character echoing.
##
##############################################################################
import readline
import sys
import threading
import time

from lib.Command import Command, command
from lib.Debug import *
from lib.Dispatch import TAB, UOW
from lib.Utility import visify
import Common

##############################################################################
## Define available imports
##############################################################################
__all__ = ['Console']

##############################################################################
## Controls
##############################################################################
_MULTI_COMMAND = False              ## Allow multiple simultaneous commands?

##############################################################################
## Import getch/putch (With internal default implementations.)
##############################################################################
try:
    raise ImportError("stdio multithreading conflict with http")
    from stdio import getch, putch
except ImportError:
    print("Warning: stdio not available")
    def getch(): ## FAKE getch (because it echos)
        return sys.stdin.read(1)

    def putch(c): ## FAKE putch (because echo already done)
        pass

##############################################################################
## Class _ConsoleTAB
##
## With only one _ConsoleTAB, commands will be processed one at a time.
## Commands can still be read, but they will be queued for processing on the
## single TAB. The _MULTI_COMMAND control variable, if True, creates a new TAB
## for each command. These commands then run in parallel.
##############################################################################
class _ConsoleTAB(TAB):
    def work(self, uow):
        work = uow.work
        if isinstance(work, str):
            if work == "":
                uow.done(None)
                return

            argv = tuple(work.split())
            work = argv[0]

            try:
                Name = command[work]
            except KeyError:
                debugf("Invalid command: '%s'" % visify(work))
                debugf("Valid commands: %s" % Command._get_list(command))
                uow.done(UOW.CC_ERROR)
                return

            try:
                uow.done(Name.run(argv))
                return
            except:
                Debug.handle_exception()

        uow.done(UOW.CC_ERROR)
        return

##############################################################################
## Class Console
##############################################################################
class Console(threading.Thread):
    __singleton = None

    def __init__(self, *args, **kwargs):
        assert Console.__singleton == None, "Only one console thread allowed"
        Console.__singleton == self

        threading.Thread.__init__(self,name="Console",*args,**kwargs)
        self.name = 'Console'
        self.tab = _ConsoleTAB()

        self.operational = True
        self.start()

    ##########################################################################
    ## This method allows for testing of different input mechanisms.
    ## Note: With Python, the last defined method is used
    def __inp(self):                ## Method[1]: sys.std.readline()
        import lib.Utility
        return lib.Utility.unbs(sys.stdin.readline())

    def __inp(self):                ## Method[2]: Character by character input
        s = ""
        while self.operational:
            c = getch()
            if c == '\b':
                if len(s) > 0:
                    putch('\b')     ## Echo character delete sequence
                    putch(' ')
                    putch('\b')
                    s = s[:-1]
            elif ord(c) == 21:      ## Ctrl-U
                L = len(s)
                for _ in range(L): putch('\b')
                for _ in range(L): putch(' ')
                for _ in range(L): putch('\b')
                s = ""
            else:
                if c == '\t': c = ' '

                putch(c) ## Echo the character
                if c == '\n':
                    break
                s += c

        return s

    def run(self):
        """Operate the Console Thread"""
        Common.add_thread(self)
        while self.operational:
            try:
                printf("%.3f >>>> " % time.time(), end="", flush=True)
                s = self.__inp().strip()
                writef("%.3f >>>> %s" % (time.time(), s))

                uow = UOW()
                uow.work = s
                tab = self.tab
                if _MULTI_COMMAND:
                    tab = _ConsoleTAB()
                Common.enqueue(tab, uow)
                time.sleep(.25)

                if s == '.quit':
                    break

            except Exception as e:
                Debug.handle_exception()
                break

        Common.del_thread(self)
        debugf("Console terminated")

    def stop(self):
        """Terminate the Console Thread"""
        self.operational = False

